add a short debug info explainer (#712)
memfault/memfault-docs: docs/reference/reference-debug-symbols.mdx

---
id: reference-debug-symbols
title: "Reference: Debug Information Symbols"
sidebar_label: "Debug Information Symbols"
---

Debug symbols allow debuggers like GDB to provide insight into a program's

> runtime by attaching symbol info to the executable. This is very useful when

Reviewer: ```suggestion
runtime by attaching debug info to the executable. This is very useful when
```

debugging issues, as you can inspect a running program with the added context of
variable names and file info. Memfault uses these symbols to build and display
backtraces captured on your device.

## Enabling Debug Info {#enabling-debug-info}

Debug info can be enabled on most C compilers by passing the `-g` compiler flag.
You can view an in-depth breakdown of the `-g` flag and some other exciting
options

> [here](https://gcc.gnu.org/onlinedocs/gcc-13.2.0/gcc/Debugging-Options.html).

Reviewer 2: ```suggestion
[in the GCC manual](https://gcc.gnu.org/onlinedocs/gcc-13.2.0/gcc/Debugging-Options.html).
```

In Linux environments, you can generate a separate archive containing all
symbols for your program and any dynamically linked programs. See our guide
[here](docs/Linux/coredumps/#upload-debugging-symbols) for Yocto Linux.

## Frequently Asked Questions {#faq}

### Does debug info make my program larger? {#faq-larger-program}

> While enabling debug will make your generated symbol file larger, in general, it

Reviewer 2: ```suggestion
While including debug symbols will make your generated ELF or binary larger, in general, it
```

Am I wrong with this correction? Not enabling debug at all won't generate a symbol file.

does not affect the size of the loaded program. This is especially true in an
MCU environment where debug info is stripped before it is loaded onto the
device.

### Does debug info make my program slower? {#faq-slower-program}

Debug symbols and optimization levels are separate configurations that do not
affect each other. Enabling debug info has no effect on execution surface of
your program.

Optimization levels can, however, make it more difficult to step through your

> program as certain names can be mangled, or functions can be inlined.

Reviewer: Maybe add a sentence what effect it can have on coredump analysis, like not being able to view certain variables, or missing call frames in the stack trace, because they got optimized out.
